scripts/prices.js

```diff
@@ -0,0 +1,73 @@
+import { getBalance } from './global';
+import { cMarket, fillCurrencySelect } from './settings';
+
+/**
+ * CoinGecko's endpoint for PIVX data, optimised for least bandwidth
+ * - No localisation, tickers, community data, developer data or sparklines
+ */
+export const COINGECKO_ENDPOINT =
+    'https://api.coingecko.com/api/v3/coins/pivx?localization=false&tickers=false&market_data=true&community_data=false&developer_data=false&sparkline=false';
+
+export class MarketSource {
+    /** The storage object for raw market data */
+    cData = {};
+
+    /** The name of the market source */
+    strName = '';
+
+    /** The customised API endpoint of the market source */
+    strEndpoint = '';
+
+    /**
+     * Ensure a market data cache exists, if not, fetch it and resume
+     */
+    async ensureCacheExists() {
+        if (!this.cData || !Object.keys(this.cData).length) await this.fetch();
+    }
+
+    /**
+     * Fetches the raw market source data
+     * @returns {Promise<object>}
+     */
+    async fetch() {
+        return (this.cData = await (await fetch(this.strEndpoint)).json());
+    }
+}
+
+export class CoinGecko extends MarketSource {
+    constructor() {
+        super();
+        this.strName = 'CoinGecko';
+        this.strEndpoint = COINGECKO_ENDPOINT;
+    }
+
+    /**
+     * Get the price in a specific display currency
+     * @param {string} strCurrency - The CoinGecko-supported display currency
+     * @return {Promise<number>}
+     */
+    async getPrice(strCurrency) {
+        await this.ensureCacheExists();
+        return this.cData.market_data.current_price[strCurrency];
+    }
+
+    /**
+     * Get a list of the supported display currencies
+     * @returns {Promise<Array<string>>} - A list of CoinGecko-supported display currencies
+     */
+    async getCurrencies() {
+        await this.ensureCacheExists();
+        return Object.keys(this.cData.market_data.current_price);
+    }
+}
+
+export async function refreshPriceDisplay() {
+    // Refresh our price data
+    await cMarket.fetch();
+
+    // Update the currency customisation menu from the selected data source
+    await fillCurrencySelect();
+
+    // Update price values
+    getBalance(true);
```

Comment:
Maybe this function can get called in getBalance instead of the other way around (makes more sense in my head)

Reply:
I did this since if we refresh market data in `getBalance()` then every balance update (which is spammed frequently during UTXO sync) would also spam the API of our data source; this way, every balance update also updates the "conversion" value without spamming the CoinGecko API.

Not sure if I misunderstood somewhere though.

Reply:
Coingecko allows 10-30 calls per minute. I think fetching the price ever 5 or 10 minutes is more then enough. After that you can do whatever you want to do with it. Even on the `getBalance()` function it uses the already fetched coingecko data.